api/condition_types.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2023 Red Hat, Inc. and/or its affiliates
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Conditions ...
+// +kubebuilder:object:generate=true
+type Conditions []Condition
+
+type ConditionType string
+
+const (
+	// RunningConditionType describes the readiness condition of a "live" resource, like the workflow application
+	RunningConditionType ConditionType = "Running"
+	// SucceedConditionType describes the readiness condition of a static resource, like a platform, a builder, a configuration, etc.
+	SucceedConditionType ConditionType = "Succeed"
+	// BuiltConditionType describes the condition of a resource that needs to be build.
+	BuiltConditionType ConditionType = "Built"
+)
+
+const (
+	WaitingForDeploymentReason  = "WaitingForDeployment"
+	DeploymentFailureReason     = "DeploymentFailure"
+	DeploymentUnavailableReason = "DeploymentIsUnavailable"
+	RedeploymentExhaustedReason = "AttemptToRedeployFailed"
+	WaitingForPlatformReason    = "WaitingForPlatform"
+	BuildFailedReason           = "BuildFailedReason"
+	WaitingForBuildReason       = "WaitingForBuild"
+)
+
+// Condition describes the common structure for conditions in our types
+// +kubebuilder:object:generate=true
+type Condition struct {
+	// Type condition for the given object
+	// +required
+	Type ConditionType `json:"type"`
+	// Status of the condition, one of True, False, Unknown.
+	// +required
+	Status v1.ConditionStatus `json:"status"`
+	// The last time this condition was updated.
+	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty"`
```

Comment:
It's not a transition, actually. It's an update. 😅

Reply:
I was referring to the meta/v1 api :)

Reply:
Oh ok! That's because we don't transition to the conditions, they are evaluated together every time. We could do it, but I don't see a benefit of doing it at the moment.

For instance, we have a `Built` condition that will be either `Unknow` or `True` in dev. We won't transition to it, since it's a static condition in the dev mode. The `LastUpdateTime` just clarifies that in both cases.